Gems/PrefabDependencyViewer/Code/Source/DirectedGraph.h

```diff
@@ -0,0 +1,231 @@
+            {
+                return m_parent;
+            }
+
+            // In the future want to be able to edit the tree.
+            void SetParent(Node* parent)
+            {
+                m_parent = parent;
+            }
+
+        private:
+            void SetMetaData(TemplateId tid, const char* source)
+            {
+                m_metaData = MetaData(tid, source);
+            }
+
+            MetaData m_metaData;
+            Node* m_parent;
+        };
+
+        class DirectedGraph
+        {
+        public:
+            DirectedGraph()
+            {
+                m_nodes = NodeSet();
+                m_children = ChildrenMap();
+                m_root = nullptr;
+            }
+
+            void AddNode(Node* node)
+            {
+                m_nodes.insert(node);
+            }
+
+            void AddChild(Node* parent, Node* child)
+            {
+                if (parent)
+                {
+                    m_children[parent].insert(child);
+                    child->SetParent(parent);
+                }
+                else
+                {
+                    m_root = child;
+                }
+            }
+
+            Node* GetRoot() const
+            {
+                return m_root;
+            }
+
+            NodeSet GetChildren(Node* parent) const
+            {
+                AZ_Assert(parent, "Nullptr can't have children");
+
+                auto iter = m_children.find(parent);
+                if (iter != m_children.end())
+                {
+                    return iter->second;
+                }
+
+                return NodeSet();
+            }
+
+            DirectedGraph(const DirectedGraph& rhs)
+            {
+                AZStd::stack<AZStd::pair<Node*, Node*>> stack;
+                stack.push(AZStd::make_pair(rhs.m_root, nullptr));
+
+                while (!stack.empty())
+                {
+                    AZStd::pair<Node*, Node*> pair = stack.top();
+                    Node* rhsNode = pair.first;
+                    Node* parent = pair.second;
+
+                    stack.pop();
+
+                    MetaData cpyMetaData = rhsNode->GetMetaData();
+                    Node* copy = aznew Node(cpyMetaData.GetTemplateId(),
+                                            cpyMetaData.GetSource(),
+                                            parent);
+
+                    AddNode(copy);
+                    AddChild(parent, copy);
+
+                    for (Node* child : rhs.GetChildren(rhsNode))
+                    {
+                        stack.push(AZStd::make_pair(child, copy));
+                    }
+                }
+            }
+
+            ~DirectedGraph()
+            {
+                auto delete_node = [](Node* node) { delete node; };
+                AZStd::for_each(m_nodes.begin(), m_nodes.end(), delete_node);
+            }
+
+            AZStd::vector<int> countNodesAtEachLevel(int& widestLevelSize) const
```

Comment:
Don't use reference-to-non-const parameters to return multiple values from a function. Use `AZStd::tuple` instead.
```suggestion
            AZStd::tuple<AZStd::vector<int>, int> countNodesAtEachLevel() const
```

Using such a function is much nicer, thanks to C++17 structured bindings:
```cpp
const auto [nodesPerLevel, widestLevelSize] = graph.countNodesAtEachLevel();
```

Reply:
Done